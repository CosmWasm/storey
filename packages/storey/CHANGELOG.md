# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.3.0] - 2024-09-11

### Added

- implement {ItemAccess, ColumnAccess}::get_or
- implement {ItemAccess, ColumnAccess}::try_get
- add Key/OwnedKey impls for more std types
- add remove to ItemAccess
- add remove entry from map

### Fixed

- fix docs inner links
- fix visibility

### Documentation

- document `BoundedIterableAccessor`
- document map key types/trait/assoc types

## [0.2.1] - 2024-06-19

### Added

- add missing IterableStorage impls

## [0.2.0] - 2024-05-09

### Added

- _(storey)_ return ix on `Column::push`
- provide typed bounds for iteration
- use single-byte keys for top-level containers

### Fixed

- _(storey)_ Column::push bounds checking

### Other

- less awkward associated type names
- rename crates to packages

## [0.1.0](https://github.com/CosmWasm/storey/releases/tag/storey-v0.1.0) - 2024-03-22

### Other

- add Cargo manifest metadata for crates.io
- implement Cw backend + encoding
- rename crates to storey
