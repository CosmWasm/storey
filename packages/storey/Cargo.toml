[package]
name = "storey"
description = "Storage abstractions for blockchains"
readme = "../../README.md"
version = "0.2.1"
edition = "2021"
rust-version = "1.65" # https://caniuse.rs/features/generic_associated_types
authors.workspace = true
license.workspace = true
repository.workspace = true
homepage.workspace = true
categories.workspace = true
keywords.workspace = true

[dependencies]
thiserror = "1"

storey-encoding.workspace = true
storey-storage.workspace = true

[dev-dependencies]
mocks = { path = "../mocks" }
