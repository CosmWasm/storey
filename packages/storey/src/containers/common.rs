#[derive(Debug, PartialEq, Eq, Clone, Copy, thiserror::Error)]
pub enum TryGetError<E> {
    #[error("item is empty")]
    Empty,
    #[error(transparent)]
    DecodeError(#[from] E),
}
