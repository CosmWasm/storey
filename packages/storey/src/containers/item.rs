use std::marker::PhantomData;

use crate::encoding::{DecodableWith, EncodableWith, Encoding};
use crate::storage::StorageBranch;
use crate::storage::{Storage, StorageMut};

use super::Storable;

/// A single item in the storage.
///
/// This simple container doesn't manage a namespace of keys, but simply stores a single
/// value under a single key.
///
/// # Example
/// ```
/// # use mocks::encoding::TestEncoding;
/// # use mocks::backend::TestStorage;
/// use storey::containers::Item;
///
/// let mut storage = TestStorage::new();
/// let item = Item::<u64, TestEncoding>::new(0);
///
/// item.access(&mut storage).set(&42).unwrap();
/// assert_eq!(item.access(&storage).get().unwrap(), Some(42));
/// ```
pub struct Item<T, E> {
    key: u8,
    phantom: PhantomData<(T, E)>,
}

impl<T, E> Item<T, E>
where
    E: Encoding,
    T: EncodableWith<E> + DecodableWith<E>,
{
    /// Create a new item with the given key.
    ///
    /// It is the responsibility of the caller to ensure that the key is unique.
    pub const fn new(key: u8) -> Self {
        Self {
            key,
            phantom: PhantomData,
        }
    }

    /// Acquire an accessor to the item.
    ///
    /// # Example
    /// ```
    /// # use mocks::encoding::TestEncoding;
    /// # use mocks::backend::TestStorage;
    /// use storey::containers::Item;
    ///
    /// // immutable accessor
    /// let storage = TestStorage::new();
    /// let item = Item::<u64, TestEncoding>::new(0);
    /// let access = item.access(&storage);
    ///
    /// // mutable accessor
    /// let mut storage = TestStorage::new();
    /// let item = Item::<u64, TestEncoding>::new(0);
    /// let mut access = item.access(&mut storage);
    pub fn access<S>(&self, storage: S) -> ItemAccess<E, T, StorageBranch<S>> {
        Self::access_impl(StorageBranch::new(storage, vec![self.key]))
    }
}

impl<T, E> Storable for Item<T, E>
where
    E: Encoding,
    T: EncodableWith<E> + DecodableWith<E>,
{
    type Accessor<S> = ItemAccess<E, T, S>;
    type Key = ();
    type KeyDecodeError = ItemKeyDecodeError;
    type Value = T;
    type ValueDecodeError = E::DecodeError;

    fn access_impl<S>(storage: S) -> ItemAccess<E, T, S> {
        ItemAccess {
            storage,
            phantom: PhantomData,
        }
    }

    fn decode_key(key: &[u8]) -> Result<(), ItemKeyDecodeError> {
        if key.is_empty() {
            Ok(())
        } else {
            Err(ItemKeyDecodeError)
        }
    }

    fn decode_value(value: &[u8]) -> Result<Self::Value, Self::ValueDecodeError> {
        T::decode(value)
    }
}

#[derive(Debug, PartialEq, Eq, Clone, Copy, thiserror::Error)]
#[error("invalid key length, expected empty key")]
pub struct ItemKeyDecodeError;

/// An accessor for an `Item`.
///
/// This type provides methods to get and set the value of the item.
pub struct ItemAccess<E, T, S> {
    storage: S,
    phantom: PhantomData<(E, T)>,
}

impl<E, T, S> ItemAccess<E, T, S>
where
    E: Encoding,
    T: EncodableWith<E> + DecodableWith<E>,
    S: Storage,
{
    /// Get the value of the item.
    ///
    /// Returns `None` if the item doesn't exist (has not been set yet).
    ///
    /// # Examples
    /// ```
    /// # use mocks::encoding::TestEncoding;
    /// # use mocks::backend::TestStorage;
    /// use storey::containers::Item;
    ///
    /// let storage = TestStorage::new();
    /// let item = Item::<u64, TestEncoding>::new(0);
    /// let access = item.access(&storage);
    ///
    /// assert_eq!(access.get().unwrap(), None);
    /// ```
    ///
    /// ```
    /// # use mocks::encoding::TestEncoding;
    /// # use mocks::backend::TestStorage;
    /// use storey::containers::Item;
    ///
    /// let mut storage = TestStorage::new();
    /// let item = Item::<u64, TestEncoding>::new(0);
    ///
    /// item.access(&mut storage).set(&42).unwrap();
    /// assert_eq!(item.access(&storage).get().unwrap(), Some(42));
    /// ```
    pub fn get(&self) -> Result<Option<T>, E::DecodeError> {
        self.storage
            .get(&[])
            .map(|bytes| T::decode(&bytes))
            .transpose()
    }
}

impl<E, T, S> ItemAccess<E, T, S>
where
    E: Encoding,
    T: EncodableWith<E> + DecodableWith<E>,
    S: StorageMut,
{
    /// Set the value of the item.
    ///
    /// # Example
    /// ```
    /// # use mocks::encoding::TestEncoding;
    /// # use mocks::backend::TestStorage;
    /// use storey::containers::Item;
    ///
    /// let mut storage = TestStorage::new();
    /// let item = Item::<u64, TestEncoding>::new(0);
    ///
    /// item.access(&mut storage).set(&42).unwrap();
    /// assert_eq!(item.access(&storage).get().unwrap(), Some(42));
    /// ```
    pub fn set(&mut self, value: &T) -> Result<(), E::EncodeError> {
        let bytes = value.encode()?;
        self.storage.set(&[], &bytes);
        Ok(())
    }

    /// Remove the value of the item.
    ///
    /// # Example
    /// ```
    /// # use mocks::encoding::TestEncoding;
    /// # use mocks::backend::TestStorage;
    /// use storey::containers::Item;
    ///
    /// let mut storage = TestStorage::new();
    /// let item = Item::<u64, TestEncoding>::new(0);
    ///
    /// item.access(&mut storage).remove();
    /// assert_eq!(item.access(&storage).get().unwrap(), None);
    /// ```
    pub fn remove(&mut self) {
        self.storage.remove(&[]);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use mocks::backend::TestStorage;
    use mocks::encoding::TestEncoding;

    #[test]
    fn basic() {
        let mut storage = TestStorage::new();

        let item0 = Item::<u64, TestEncoding>::new(0);
        item0.access(&mut storage).set(&42).unwrap();

        let item1 = Item::<u64, TestEncoding>::new(1);
        let access1 = item1.access(&storage);

        assert_eq!(item0.access(&storage).get().unwrap(), Some(42));
        assert_eq!(storage.get(&[0]), Some(42u64.to_le_bytes().to_vec()));
        assert_eq!(access1.get().unwrap(), None);
        assert_eq!(storage.get(&[1]), None);
    }
}
