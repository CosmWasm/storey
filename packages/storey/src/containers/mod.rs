//! This module contains both the traits for implementing collections/containers, as well as a
//! few fundamental collections/containers themselves.

mod column;
pub mod common;
mod item;
pub mod map;
#[cfg(test)]
mod test_utils;

use std::{marker::PhantomData, ops::Bound};

use storey_storage::RevIterableStorage;

use crate::storage::IterableStorage;

pub use storey_macros::router;

pub use column::{Column, ColumnAccess};
pub use item::{Item, ItemAccess};
pub use map::{Map, MapAccess};

/// The fundamental trait every collection/container should implement.
pub trait Storable {
    type Kind: StorableKind;

    /// The accessor type for this collection/container. An accessor is a type that provides
    /// methods for reading and writing to the collection/container and encapsulates the
    /// specific [`Storage`] type used (the `S` type parameter here).
    ///
    /// [`Storage`]: crate::storage::Storage
    type Accessor<S>;

    /// Create an accessor for this collection/container, given a [`Storage`] implementation.
    ///
    /// [`Storage`]: crate::storage::Storage
    fn access_impl<S>(storage: S) -> Self::Accessor<S>;
}

pub trait IterableStorable: Storable {
    /// The Key type for this collection/container. This is the type that will be used in
    /// key iteration.
    ///
    /// For composable collections this is the "full" key, e.g. for [`Map`]
    /// this is a tuple of the key and the sub-key.
    ///
    /// Containers that store one item and don't manage subkeys should use the `()` type here.
    type Key;

    /// The error type for decoding keys.
    type KeyDecodeError;

    /// The Value type for this collection/container. This is the type that will be used for
    /// value iteration.
    type Value;

    /// The error type for decoding values.
    type ValueDecodeError;

    /// Decode a key from a byte slice.
    ///
    /// This method is used in key iteration to provide a typed key rather than raw bytes
    /// to the user.
    fn decode_key(key: &[u8]) -> Result<Self::Key, Self::KeyDecodeError>;

    /// Decode a value from a byte slice.
    ///
    /// This method is used in value iteration to provide a typed value rather than raw bytes
    /// to the user.
    fn decode_value(value: &[u8]) -> Result<Self::Value, Self::ValueDecodeError>;
}

/// A key-value pair decoding error.
#[derive(Debug, PartialEq, thiserror::Error)]
pub enum KVDecodeError<K, V> {
    #[error("failed to decode key: {0}")]
    Key(K),
    #[error("failed to decode value: {0}")]
    Value(V),
}

impl<K: std::fmt::Display, V: std::fmt::Display> crate::error::StoreyError for KVDecodeError<K, V> {}

/// A trait for collection accessors (see [`Storable::Accessor`]) that provide iteration over
/// their contents.
pub trait IterableAccessor: Sized {
    /// The [`Storable`] type this accessor is associated with.
    type Storable: Storable;

    /// The [`Storage`] type this accessor is associated with.
    ///
    /// [`Storage`]: crate::storage::Storage
    type Storage: IterableStorage;

    /// Get a reference to the storage this accessor is associated with.
    fn storage(&self) -> &Self::Storage;

    /// Iterate over key-value pairs in this collection.
    fn pairs(
        &self,
    ) -> StorableIter<Self::Storable, <Self::Storage as IterableStorage>::PairsIterator<'_>> {
        StorableIter {
            inner: self.storage().pairs(Bound::Unbounded, Bound::Unbounded),
            phantom: PhantomData,
        }
    }

    /// Iterate over keys in this collection.
    fn keys(
        &self,
    ) -> StorableKeys<Self::Storable, <Self::Storage as IterableStorage>::KeysIterator<'_>> {
        StorableKeys {
            inner: self.storage().keys(Bound::Unbounded, Bound::Unbounded),
            phantom: PhantomData,
        }
    }

    /// Iterate over values in this collection.
    fn values(
        &self,
    ) -> StorableValues<Self::Storable, <Self::Storage as IterableStorage>::ValuesIterator<'_>>
    {
        StorableValues {
            inner: self.storage().values(Bound::Unbounded, Bound::Unbounded),
            phantom: PhantomData,
        }
    }
}

pub trait RevIterableAccessor
where
    Self: IterableAccessor,
    Self::Storage: RevIterableStorage,
{
    /// Iterate over key-value pairs in this collection in reverse order.
    fn rev_pairs(
        &self,
    ) -> StorableIter<Self::Storable, <Self::Storage as RevIterableStorage>::RevPairsIterator<'_>>
    {
        StorableIter {
            inner: self.storage().rev_pairs(Bound::Unbounded, Bound::Unbounded),
            phantom: PhantomData,
        }
    }

    /// Iterate over keys in this collection in reverse order.
    fn rev_keys(
        &self,
    ) -> StorableKeys<Self::Storable, <Self::Storage as RevIterableStorage>::RevKeysIterator<'_>>
    {
        StorableKeys {
            inner: self.storage().rev_keys(Bound::Unbounded, Bound::Unbounded),
            phantom: PhantomData,
        }
    }

    /// Iterate over values in this collection in reverse order.
    fn rev_values(
        &self,
    ) -> StorableValues<Self::Storable, <Self::Storage as RevIterableStorage>::RevValuesIterator<'_>>
    {
        StorableValues {
            inner: self
                .storage()
                .rev_values(Bound::Unbounded, Bound::Unbounded),
            phantom: PhantomData,
        }
    }
}

impl<I> RevIterableAccessor for I
where
    I: IterableAccessor,
    I::Storage: RevIterableStorage,
{
}

/// This trait extends [`IterableAccessor`] with methods for bounded iteration. Not every
/// iterable collection supports it, so this trait is separate.
///
/// Bounded iteration allows the user to specify a start and end bound for the iteration.
///
/// # Why not always support bounded iteration?
///
/// The reason bounded iteration isn't always supported is that it may not be possible to
/// efficiently implement it. For example, a `Map<String, Map<u8, _>` must length-prefix
/// the string keys when mapping them to the byte keys used in the storage backend. Otherwise,
/// we'd not know where the string key ends and the nested `u8` key begins.
///
/// The length prefix will interfere with the ordering of the keys so that it's no longer
/// lexicographical. It's deterministic, but rather confusing and unlikely to be useful. This
/// in turn means the entries found between two string keys may not be the expected ones.
pub trait BoundedIterableAccessor: IterableAccessor {
    /// Iterate over key-value pairs in this collection, respecting the given bounds.
    ///
    /// Either end of the range can be unbounded, inclusive, or exclusive. See [`Bound`] for more.
    fn bounded_pairs<B>(
        &self,
        start: Bound<B>,
        end: Bound<B>,
    ) -> StorableIter<Self::Storable, <Self::Storage as IterableStorage>::PairsIterator<'_>>
    where
        B: BoundFor<Self::Storable>,
    {
        let start = start.map(|b| b.into_bytes());
        let end = end.map(|b| b.into_bytes());

        StorableIter {
            inner: self.storage().pairs(
                start.as_ref().map(|b| b.as_slice()),
                end.as_ref().map(|b| b.as_slice()),
            ),
            phantom: PhantomData,
        }
    }

    /// Iterate over keys in this collection, respecting the given bounds.
    ///
    /// Either end of the range can be unbounded, inclusive, or exclusive. See [`Bound`] for more.
    fn bounded_keys<B>(
        &self,
        start: Bound<B>,
        end: Bound<B>,
    ) -> StorableKeys<Self::Storable, <Self::Storage as IterableStorage>::KeysIterator<'_>>
    where
        B: BoundFor<Self::Storable>,
    {
        let start = start.map(|b| b.into_bytes());
        let end = end.map(|b| b.into_bytes());

        StorableKeys {
            inner: self.storage().keys(
                start.as_ref().map(|b| b.as_slice()),
                end.as_ref().map(|b| b.as_slice()),
            ),
            phantom: PhantomData,
        }
    }

    /// Iterate over values in this collection, respecting the given bounds.
    ///
    /// Either end of the range can be unbounded, inclusive, or exclusive. See [`Bound`] for more.
    fn bounded_values<B>(
        &self,
        start: Bound<B>,
        end: Bound<B>,
    ) -> StorableValues<Self::Storable, <Self::Storage as IterableStorage>::ValuesIterator<'_>>
    where
        B: BoundFor<Self::Storable>,
    {
        let start = start.map(|b| b.into_bytes());
        let end = end.map(|b| b.into_bytes());

        StorableValues {
            inner: self.storage().values(
                start.as_ref().map(|b| b.as_slice()),
                end.as_ref().map(|b| b.as_slice()),
            ),
            phantom: PhantomData,
        }
    }
}

/// This trait extends [`BoundedIterableAccessor`] with methods for bounded reverse iteration.
/// Not every iterable collection supports it, so this trait is separate.
///
/// Bounded reverse iteration allows the user to specify a start and end bound for the iteration,
/// but in reverse order.
///
/// # Why not always support bounded reverse iteration?
///
/// The same reasons as for [bounded iteration](BoundedIterableAccessor) apply.
pub trait BoundedRevIterableAccessor
where
    Self: BoundedIterableAccessor,
    Self::Storage: RevIterableStorage,
{
    /// Iterate over key-value pairs in this collection in reverse order, respecting the given bounds.
    ///
    /// Either end of the range can be unbounded, inclusive, or exclusive. See [`Bound`] for more.
    fn bounded_rev_pairs<B>(
        &self,
        start: Bound<B>,
        end: Bound<B>,
    ) -> StorableIter<Self::Storable, <Self::Storage as RevIterableStorage>::RevPairsIterator<'_>>
    where
        B: BoundFor<Self::Storable>,
    {
        let start = start.map(|b| b.into_bytes());
        let end = end.map(|b| b.into_bytes());

        StorableIter {
            inner: self.storage().rev_pairs(
                start.as_ref().map(|b| b.as_slice()),
                end.as_ref().map(|b| b.as_slice()),
            ),
            phantom: PhantomData,
        }
    }

    /// Iterate over keys in this collection in reverse order, respecting the given bounds.
    ///
    /// Either end of the range can be unbounded, inclusive, or exclusive. See [`Bound`] for more.
    fn bounded_rev_keys<B>(
        &self,
        start: Bound<B>,
        end: Bound<B>,
    ) -> StorableKeys<Self::Storable, <Self::Storage as RevIterableStorage>::RevKeysIterator<'_>>
    where
        B: BoundFor<Self::Storable>,
    {
        let start = start.map(|b| b.into_bytes());
        let end = end.map(|b| b.into_bytes());

        StorableKeys {
            inner: self.storage().rev_keys(
                start.as_ref().map(|b| b.as_slice()),
                end.as_ref().map(|b| b.as_slice()),
            ),
            phantom: PhantomData,
        }
    }

    /// Iterate over values in this collection in reverse order, respecting the given bounds.
    ///
    /// Either end of the range can be unbounded, inclusive, or exclusive. See [`Bound`] for more.
    fn bounded_rev_values<B>(
        &self,
        start: Bound<B>,
        end: Bound<B>,
    ) -> StorableValues<Self::Storable, <Self::Storage as RevIterableStorage>::RevValuesIterator<'_>>
    where
        B: BoundFor<Self::Storable>,
    {
        let start = start.map(|b| b.into_bytes());
        let end = end.map(|b| b.into_bytes());

        StorableValues {
            inner: self.storage().rev_values(
                start.as_ref().map(|b| b.as_slice()),
                end.as_ref().map(|b| b.as_slice()),
            ),
            phantom: PhantomData,
        }
    }
}

impl<I> BoundedRevIterableAccessor for I
where
    I: BoundedIterableAccessor,
    I::Storage: RevIterableStorage,
{
}

/// A type that can be used as bounds for iteration over a given collection.
///
/// As an example, a collection `Foo` with string-y keys can accept both `String` and
/// `&str` bounds by providing these impls:
/// - `impl BoundFor<Foo> for &str`
/// - `impl BoundFor<Foo> for String`
pub trait BoundFor<T> {
    fn into_bytes(self) -> Vec<u8>;
}

/// The iterator over key-value pairs in a collection.
pub struct StorableIter<S, I> {
    inner: I,
    phantom: PhantomData<S>,
}

impl<S, I> Iterator for StorableIter<S, I>
where
    S: IterableStorable,
    I: Iterator<Item = (Vec<u8>, Vec<u8>)>,
{
    type Item = Result<(S::Key, S::Value), KVDecodeError<S::KeyDecodeError, S::ValueDecodeError>>;

    fn next(&mut self) -> Option<Self::Item> {
        self.inner.next().map(|(k, v)| -> Self::Item {
            match (S::decode_key(&k), S::decode_value(&v)) {
                (Err(e), _) => Err(KVDecodeError::Key(e)),
                (_, Err(e)) => Err(KVDecodeError::Value(e)),
                (Ok(k), Ok(v)) => Ok((k, v)),
            }
        })
    }
}

/// The iterator over keys in a collection.
pub struct StorableKeys<S, I> {
    inner: I,
    phantom: PhantomData<S>,
}

impl<S, I> Iterator for StorableKeys<S, I>
where
    S: IterableStorable,
    I: Iterator<Item = Vec<u8>>,
{
    type Item = Result<S::Key, S::KeyDecodeError>;

    fn next(&mut self) -> Option<Self::Item> {
        self.inner.next().map(|k| S::decode_key(&k))
    }
}

/// The iterator over values in a collection.
pub struct StorableValues<S, I> {
    inner: I,
    phantom: PhantomData<S>,
}

impl<S, I> Iterator for StorableValues<S, I>
where
    S: IterableStorable,
    I: Iterator<Item = Vec<u8>>,
{
    type Item = Result<S::Value, S::ValueDecodeError>;

    fn next(&mut self) -> Option<Self::Item> {
        self.inner.next().map(|v| S::decode_value(&v))
    }
}

/// The kind of a storable.
///
/// This is used to differentiate between terminal and non-terminal storables.
/// See also: [`Terminal`] and [`NonTerminal`].
///
/// This trait is [sealed](https://rust-lang.github.io/api-guidelines/future-proofing.html#sealed-traits-protect-against-downstream-implementations-c-sealed)
/// and cannot be implemented outside of this crate.
pub trait StorableKind: sealed::StorableKindSeal {}

/// A terminal [`Storable`] kind. A terminal storable doesn't manage any subkeys,
/// and is the end of the line in a composable collection.
///
/// An example of a terminal storable is [`Item`], but not [`Column`] or [`Map`].
pub struct Terminal;

/// A non-terminal [`Storable`] kind. A non-terminal storable manages subkeys.
///
/// Some examples of non-terminal storables are [`Column`] and [`Map`].
pub struct NonTerminal;

impl StorableKind for Terminal {}
impl StorableKind for NonTerminal {}

mod sealed {
    pub trait StorableKindSeal {}

    impl StorableKindSeal for super::Terminal {}
    impl StorableKindSeal for super::NonTerminal {}
}
