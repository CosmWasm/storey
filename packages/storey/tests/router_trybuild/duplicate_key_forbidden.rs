use mocks::encoding::TestEncoding;
use storey::containers::{router, Item, Map};

router! (
    router Foo {
        2 -> a: Item<u64, TestEncoding>,
        1 -> b: Map<String, Item<u64, TestEncoding>>,
        2 -> c: Item<u64, TestEncoding>,
    }
);

fn main() {}
