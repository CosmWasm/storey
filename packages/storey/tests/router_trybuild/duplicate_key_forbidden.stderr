error: Duplicate field key: 2
 --> tests/router_trybuild/duplicate_key_forbidden.rs:8:9
  |
8 |         2 -> c: Item<u64, TestEncoding>,
  |         ^

error: Duplicate field key: 2
 --> tests/router_trybuild/duplicate_key_forbidden.rs:9:9
  |
9 |         2 -> d: Item<u64, TestEncoding>,
  |         ^

warning: unused import: `mocks::encoding::TestEncoding`
 --> tests/router_trybuild/duplicate_key_forbidden.rs:1:5
  |
1 | use mocks::encoding::TestEncoding;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: unused imports: `Item` and `Map`
 --> tests/router_trybuild/duplicate_key_forbidden.rs:2:34
  |
2 | use storey::containers::{router, Item, Map};
  |                                  ^^^^  ^^^
